keps/sig-api-machinery/0006-apply.md

```diff
@@ -194,8 +195,27 @@ atomic. That can be specified with `// +mapType=atomic` or `//
 +structType=atomic` respectively. They map to the same openapi extension:
 `"x-kubernetes-map-type": "atomic"`.
 
-#### Kubectl
-##### Server-side Apply
+#### Clients
+
+##### Apply Status Header
+
+Client side apply in kubectl provides the user with information if the applied config resulted in any changes
+on the apiserver.
+
+To provided the same information with server-side apply, a dedicated header is set by the apiserver
+if the request caused changes to the existing resource.
+
+```txt
+X-Kubernetes-Apply-Status: "modified" || "unmodified"
```

Comment:
Would you mind describing all possible values and what they mean exactly? What does it mean for the resource to be modified. I'm also assuming we only send this annotation on apply? Can it be missing? What does missing mean? Old versions don't send it so I'm guessing it must be allowed.

Reply:
I am split between:
- only send the header, with one value, if apply changed (i had it `X-Kubernetes-Applied: "true"`)
- send "modified" for applies that updated the managedFields, "unmodified" for ones that didn't (ignoring stripped/removed fields). Missing the header would be "unknown" for the client and most likely indicates the server is old.

We could set the header for other operations as well. I don't really have an opinion on this right now. It would need a different name then.

Reply:
I suggest "Changes-Made" and "Changes-Unneeded" as the two string values. (just saying "unmodified" could be confused for an error condition, I think.)

Less important, consider instead of "X-Kubernetes-Apply-Success-Type" as the header name unless you intend to add in other values (e.g., something about a conflict).

Reply:
Yeah, i wouldn't put conflict in there. it is already clear from the error (and status code i think).

Don't have a strong opinion on the header name aside from that.

Reply:
On my comment before, i tend to stick with some kind of changed/unchanged so that header absence can not mean two things (unchanged|unknown), but only that the server does not know the header.

@lavalamp what do you think about adding the header for non-apply as well?

Reply:
I think it's OK to name the header such that it'd make sense on other verbs, but let's not actually implement it for anything but APPLY for the moment.

Reply:
I think we should say this is an optional header and clients shouldn't deduce anything in particular from its absence.